components/wmtk_components/embedding/internal/Embedding.cpp

```diff
@@ -0,0 +1,152 @@
+    Eigen::MatrixXd& m_vertices_,
+    EmbeddingOptions& options_,
+    double m_blank_rate_
+    // double m_resolute_area_)
+    )
+    : m_edges(m_edges_)
+    , m_vertices(m_vertices_)
+    , options(options_)
+    , m_blank_rate(m_blank_rate_)
+//, m_resolute_area(m_resolute_area_)
+{
+    m_marked_vertices.reserve(m_vertices.rows());
+    m_marked_edges.reserve(m_edges.rows());
+
+    for (size_t i = 0; i < m_vertices.rows(); ++i) {
+        m_marked_vertices.emplace_back(i);
+    }
+    for (size_t i = 0; i < m_edges.rows(); ++i) {
+        m_marked_edges.emplace_back(std::pair<long, long>(m_edges(i, 0), m_edges(i, 1)));
+    }
+}
+void Embedding::compute_bounding_value(double& max_x, double& max_y, double& min_x, double& min_y)
+{
+    max_x = max_y = std::numeric_limits<double>::lowest();
+    min_x = min_y = std::numeric_limits<double>::max();
+    for (size_t i = 0; i < m_vertices.rows(); ++i) {
+        max_x = std::max(max_x, m_vertices(i, 0));
+        max_y = std::max(max_y, m_vertices(i, 1));
+        min_x = std::min(min_x, m_vertices(i, 0));
+        min_y = std::min(min_y, m_vertices(i, 1));
+    }
+    const double blank_region_length_x = (max_x - min_x) * m_blank_rate;
+    const double blank_region_length_y = (max_y - min_y) * m_blank_rate;
+    max_x += blank_region_length_x;
+    min_x -= blank_region_length_x;
+    max_y += blank_region_length_y;
+    min_y -= blank_region_length_y;
+}
+void Embedding::process()
+{
+    // find boundingbox
+    double max_x, min_x, max_y, min_y;
+    compute_bounding_value(max_x, max_y, min_x, min_y);
+
+    Eigen::MatrixXd bounding_box_vertices(4, 2);
+    Eigen::Matrix<long, -1, -1> bounding_box_edges(4, 2);
+    bounding_box_vertices << max_x, max_y, min_x, max_y, min_x, min_y, max_x, min_y;
+    bounding_box_edges << m_marked_vertices.size(), m_marked_vertices.size() + 1,
+        m_marked_vertices.size() + 1, m_marked_vertices.size() + 2, m_marked_vertices.size() + 2,
+        m_marked_vertices.size() + 3, m_marked_vertices.size() + 3, m_marked_vertices.size();
+
+    // NOTE: make sure give the matrix right size before assign a value to them
+    Eigen::MatrixXd temp_vertices(m_vertices.rows() + bounding_box_vertices.rows(), 2);
+    Eigen::Matrix<long, -1, -1> temp_edges(m_edges.rows() + bounding_box_edges.rows(), 2);
+    temp_vertices << m_vertices, bounding_box_vertices;
+    temp_edges << m_edges, bounding_box_edges;
+
+    // this is for triangulate function, should be outside of the bounding box.
+    Eigen::MatrixXd outside_vertex;
+    outside_vertex.resize(1, 2);
+    outside_vertex << max_x + 1.0, max_y + 1.0;
+
+    // for (int i = 0; i < m_vertices.rows(); i++) {
+    //     spdlog::info("{} {}", m_vertices(i, 0), m_vertices(i, 1));
+    // }
+
+    // for (int i = 0; i < m_edges.rows(); i++) {
+    //     spdlog::info("{} {}", m_edges(i, 0), m_edges(i, 1));
+    // }
+
+    // the fourth parameter is resolution, we should discuss it maybe.

```

Comment:
The embedding component should do as least as possible, i.e. the worst resolution that is still valid.

Reply:
Okay, then I will set the resolution constant as "a0.1q", which means
a0.1 means that the area of each triangle should be not greater than 0.1
q means that no angles will be smaller than 20 degrees

Reply:
If possible, remove this completely or set it to the minimum value possible. We just want a valid solution from the embedding, not necessarily a good or even reasonable one.